lambda/matching.ml

```diff
@@ -3650,40 +3675,40 @@ let do_for_multiple_match ~scopes loc paraml pat_act_list partial =
       } )
   in
   try
-    try
-      (* Once for checking that compilation is possible *)
-      let next, nexts =
-        split_and_precompile ~arg_id:None ~arg_lambda:arg pm1
-      in
-      let size = List.length paraml
-      and idl = List.map (fun _ -> Ident.create_local "*match*") paraml in
-      let args = List.map (fun id -> (Lvar id, Alias)) idl in
-      let flat_next = flatten_precompiled size args next
-      and flat_nexts =
-        List.map (fun (e, pm) -> (e, flatten_precompiled size args pm)) nexts
-      in
-      let lam, total =
-        comp_match_handlers (compile_flattened ~scopes repr) partial
-          (Context.start size) flat_next flat_nexts
-      in
-      List.fold_right2 (bind Strict) idl paraml
-        ( match partial with
+    match split_and_precompile ~arg_id:None ~arg_lambda:arg pm1 with
+    | exception Cannot_flatten ->
+        (* One pattern binds the whole tuple, flattening is not possible.
+           We need to allocate the scrutinee. *)
+        let lambda, total =
+          compile_match ~scopes None partial (Context.start 1) pm1 in
+        begin match partial with
         | Partial ->
-            check_total total lam raise_num (partial_function ~scopes loc)
+            check_total total lambda raise_num (partial_function ~scopes loc)
         | Total ->
             assert (Jumps.is_empty total);
-            lam
-        )
-    with Cannot_flatten -> (
-      let lambda, total =
-        compile_match ~scopes None partial (Context.start 1) pm1 in
-      match partial with
-      | Partial ->
-          check_total total lambda raise_num (partial_function ~scopes loc)
-      | Total ->
-          assert (Jumps.is_empty total);
-          lambda
-    )
+            lambda
+        end
+    | next, nexts ->
+        let size = List.length paraml
+        and idl = List.map (fun _ -> Ident.create_local "*match*") paraml in
+        let args = List.map (fun id -> (Lvar id, Alias)) idl in
+        (* This will not raise, since the PM is well typed. *)
```

Comment:
@trefis on rereading the PR, I find this comment a bit obscure. Maybe you could explain slightly more or at least refer to the other comment in the `flatten_...` function.

Reply:
I agree with you that the comment is not actually helping much as is.
Btw, `flatten_precompiled` actually doesn't raise anymore (it fatal errors), so perhaps we just don't need the comment at all?

Reply:
I removed the comment. I think there is more to do on this piece of code, but this PR is already a good improvement.